### DISCUSSION TITLE
Would it make sense to add a move constructor to the context class (C++)?

### DISCUSSION THREAD

**Discussion by jparsert** [Z3Prover/z3#6557] [General]
The context as far as I understand manages objects/configurations etc. In an implementation of mine I am writing a wrapper class that contains context/solver/terms etc. Now this wrapper is has implicitly deleted copy/move constructors because one of its members, namely `context` has explicitly deleted copy/move constructors (line 176 of `z3++.h`).

Hence, I was wondering if it makes sense to have multiple contexts in the same application. In my case I consider multiple SMT problems simultaneously so I create a new context/solver for each problem.

And if it makes sense to have multiple contexts, why are copy and move constructors deleted and wouldn't it make sense to at least provide a move-constructor?

**Comment by NikolajBjorner** [comment]
Someone added move constructors to expr and it broke the API. because it wasn't playing correctly with reference counts in the ways expressions were used elsewhere. 
You can use multiple contexts if you run solvers in parallel (different context per thread) or want to control memory locality.

**Reply by jparsert** [reply]
Ah, I see. I was afraid this might be the case. So I guess for now the best thing is to just pass the wrapper by reference or pointer and using emplace_back when using containers.
